meson: Take git (almost) completely out of builds

Finii opened the pull request:
This is almost ready.

- The API version is taken from the new `APIVERSION` file
- The library version is taken from the new `LIBNO` file, maybe
- The LIBNO file has the same shell-script contents `LIBNO=x.y.z` as usual
- The LIBNO file is only used if `-Ddeb-vers-tag` is set/changed by the user (typically done for Doocs builds)
- The `deb-vers-tag` is only used to determine if LIBNO is used :unamused:, backwards compatible
- Automatic patchlevel builds/packages are not possible anymore and we need to fall back to `makeDdeb -l` :disappointed: 
- The general way to packetize is unchanged (which is also unnecessarily complicated)

At the moment the `debian/changelog` is still autogenerated using the `deb-vers-tag`. But that can lead to problems if the `APIVERSION` / `LIBNO` file get out of sync of the git tags. Will examine that later/tomorrow.

### Old

```
Message: LIBGUL_API_VERSION       2.12
Message: GIT FULL API VERSION    v2.12.1-22-gaf0b55b
Message: GIT VERSION TAG         DOOCSVERSION_24_12_1
Message: Debian package names    doocs-gul17-24-12-1, dev-doocs-gul17
Message: Debian package version  24.12.1.p25
Message: Shared library version  24.12.1
```

### New

```
Message: API VERSION             2.12
Message: FULL API VERSION        2.12.1
Message: LIB VERSION             24.12.1
Message: Debian package names    doocs-gul17-24-12-1, dev-doocs-gul17
Message: Debian package version  24.12.1
Message: Shared library version  24.12.1
```

_FULL API VERSION_ = file `APIVERSION`
_LIB VERSION_ = Full API version; except when `deb-vers-tag` is changed than file `LIBNO`

Ping @alt-graph @ohensler @soerengrunewald

Finii commented:
💭 Tests fail? Who cares ;-)

> Do we actually need an API version?

Well, this PR does not change the general ideas and setup. You still can, and by default do crate an ordinary library package with API version as library version and the version NOT in the name, so that 'centralistic' library updates are possible. What everyone does except doocs ;-)

Only if you set `deb-vers-pack` _"Debian package name will contain patchlevel"_ will the "Doocs version scheme" be used and LIBNO contains the version to use.

It can even be more simple (for some definitions of simple), for example `libslx` packaging of servers is different in the `debuild` / `makeDdeb` call. You do that in the source_root and not in some setup build stuff. But maybe that is also personal preference. There you are in the source root on the command line, set the (annotated) version tag and then call debuild and the new package is created.

...

But you are right we do not need an `APIVERSION` file at all to do that, and it is only read once on Meson configure to set the project version. We cal also set the project version directly... changing.

Finii commented:
> Do we actually need an API version? I am not so sure that anyone really used our version number in the "semantic versioning" sense.

Read your comment again after some sleep ;)

Yes, maybe we should completely drop the 'ordinary versioning scheme', which would also resolve the strange dependencies with taskolib.

I'll swap that out, and we can ask Tim if he is fine with a `meson.build` based approach (I remember giving him the `meson rewrite` examples) or if LIBNO is better which is quickly refittet.

Also will remove the `changelog`, which is based on git at the moment but can come out of sync if the git tags are not checked and the changelog is transitional and also never looked at. Then we can also live with a fake changelog file.

Finii commented:
> but we could invent something like `24.12.1a`.

Well, that is already automated, see image below. If I would package now it will add the number of commits since the latest version.

But this has probably never been used with GUL, but this is how libslx is used always. I guess the git commit based versioning was not understood, and people always thought about `makeDdeb -l 2` or other long obsolete methods to create an additional package with a "new" name. This `-xenial1` was a thing in year 2000, but nowadays noone uses that anymore.
But we are tied to this bloody platform name (which is not needed nowadays anymore) because `makeDdeb` "secretly" changes out changelog, commits the change, packages, and then `reset --hard HEAD^` :sob: 

That is "in conflict" with ... just do an ordinary `debuild` or `makeDdeb -I` eben, and the new version will be autogenerated.

Well. I know what I would prefer, but obviously that has not taken off.

So, and that would be a question. To have this automagic number version thing that counts the commits since last tagged version, we need to depend on git (at least for the changelog). Do we want that, or do we continue with the 20th century approach and `-xenial2` style manual tinkery.

![image](https://github.com/user-attachments/assets/e2e45bea-b62e-47d8-8027-6889c07895cf)

Finii commented:
Thinking about all the consequences of a "simple" setup (without git), I am not so sure we really want that.

In other places "make simple things simple" and "make it hard to use wrong" is sooo important, and here with the package building we step back from a system that could almost not at all be used wrong and is simple to use (with complex code), we step back to something that is simple to implement but that does not guarantee and check anything and "dann muss man halt beim Packaging aufpassen" :-(

And we DO have (thank god) the git tag based versioning already is some other libs and a lot servers, and I believe that is the way to go. To not have the information stored in two different "databases" independently.
One could think it the approach command-wise is good. If the coupling of Debian-packaging preparation and Meson is so good (probably not). But that is different from lets throw out git-use for building/packaging.

What I think what we can simplify here and probably SHOULD, is just removing the possibility to deviate from the doocs way of versions. If we remove that it will already be simpler and more like in the clientlib for example.

Had headache the whole day. Maybe this is all wasted, but then, better wasted and we have a good solution than to continue with something we do not really want.

Please share thoughts ;-)

alt-graph commented:
> But we are tied to this bloody platform name (which is not needed nowadays anymore) because makeDdeb "secretly" changes our changelog, commits the change, packages, and then reset --hard HEAD^ 😭
>That is "in conflict" with ... just do an ordinary debuild or makeDdeb -I even, and the new version will be autogenerated.

Just thinking out loud... I have never looked into `makeDdeb`, but can't we change its behavior for the case where there is no `LIBNO` file, but just a `meson.build`? I won't object if `makeDdep` depends on a git repo and git tags and whatever, but having the library build script itself interacting with git always leaves a bad taste in my mouth – as in "a piece of code should be buildable from a tarball". 🤔

Finii commented:
The code was (at least should have been) buildable from a tarball.
Using git is about the packaging. And about maintenance.

Usually project and packaging are two (or more) distinct repos; Doocs typically combines project and debian packaging into one repo for convenience and because project and package maintainer is the same one person.

> I won't object if makeDdep depends on a git repo and git tags and whatever

Now that you mention that, I remember that I implemented all that stuff into `makeDdeb` already some time ago. But as I have a look now, that is only for servers and not for libraries.

Unlike the debian packaging setup we have here now (which everyone can use with the Ubuntu provided tools and is imho quite straightforward), `makeDdeb` uses sed and awk galore and sub-scripts where some have to be in-repo and some are dependencies that need to be installed. That is really a mess, esp compared with what we have added here in the meson build files.

So the idea behind all the packaging stuff I put into the meson files, that is to avoid the need of any of these non-standard tools.

In contrast my goal was to have everything out in the open, in the repo, not depending on some script tools that furthermore depend on other scripts in various packages. Instead of a not-documented way to create the packages (that depend on some person maintaining buildhosts in a certain way which is also undocumented) I wanted a clean way to create the packages, fully documented and fully understandable within that repo.

Those tools all originate from pre year 2000 and how Debian packages have been built back then. They implement kludges and deviation from "the standard way of doing" things that drives tears to my eyes (*).
(Solely) relying on the standard tools makes sure we build stuff how it is currently expected to be build. We can have almost empty debian files, which allow painless upgrades to new debian packaging, which is constantly evolving.

Not only "one should be able to build the project from a tarball" but also "one should be able to package the project from its packaging-repo without any unusual dependencies".

> for the case where there is no LIBNO file, but just a meson.build?

In principle, of course. But `makeDdeb` is already an unmaintainable mess of historically grown code.
It would be far easier to create a new tool `ninjaDdeb` or `mesonDdeb` that does not need to know about Makefiles, about `check_rules` scripts or `make-doocspkg`.

(*) WHY WHY are the development packages called `dev-doocs-...` instead of `doocs-...-dev` which is the standard for debian packages? WHY put the OS nickname on patch release package names (`-xenial1`) instead of a more sane `-p1`? WHY don't we have `.ddeb` packages but instead leave the debug symbols in the executables? :sob: This all feels so 1990.

---

I rewrote this now several times. I do not know where we aim at, I just know that my frustration with the doocs idiosyncrasies is rather high and I want to get rid of as many as possible, The doocs packaging is a quagmire.

...

Rambling along here.
I believe it would be good to rip the packaging out of the repos. That is also the way we do it with `libslx`.
The later needs to be used as submodule which has pros and cons; and this has been discussed by Soeren and me several times, how it can be solved, and one way is indeed to have some helper tool installed as separate package (like `makeDdeb`).

Did you ever looked into a `meson.build` of a server using libslx? I believe that it is close to "my ideal" as it is concise, simple and clear. Whatever can be deduced from somewhere else IS deduced. All the debian files that can be autogenerated are generated on the fly or modified in a clean way and not with `sed` and hoping it will find the correct string to replace.

```python
# Build specification

project('vat590', [ 'cpp', 'c' ],
    default_options : [ 'cpp_std=c++17' ],
    meson_version : '>=0.53.2'.
)
watchdogname = 'VAT'
subdir('libslx')

src = [
   'eq_vat590.cc',
   'sensors.c',
   'vat590_rpc_server.cc'
]

executable(server_exe, src + [ gitversion_cc ],
           include_directories : incdir,
           dependencies : [ doocs_dep ],
           link_with : libslx,
           install : true,
           install_dir : installdir,
)
```

Hmmm.

At first I will finish this here maybe, and have it create nonsensical changelogs if there are conflicts, but have the package created based on non-git information.

Finii commented:
Apropos `makeDdeb` ... I remember we had a discussion if the main server source files needs to be called `..._rpc_server.cc` or just `main.cc` which you preferred iirc.

OF COURSE that breaks building with the old build helpers, here `make-doocspkg` which is sometimes called from `makeDdeb` :grimacing: 

![image](https://github.com/user-attachments/assets/4948705f-804d-4067-8221-91c78b5e8fd6)

Of course this is only true with Makefile based projects, but then, they are not forbidden. These tools need a lot of stuff to be in some non-documented but expected state, it is so extremely brittle.

Just saw that when working on https://mcs-gitlab.desy.de/doocs/doocs-examples/clean_template/-/issues/2

Finii commented:
`makeDdeb` and servers, there it seems to work already, also undocumented of course :grimacing: 

![image](https://github.com/user-attachments/assets/acd5a097-9e09-48ca-bb2b-7af92d571e26)

Maybe we can really expand that to libs.

alt-graph reviewed (approved):
Thanks a lot, @Finii! I see this is still a draft, but it makes the build script already much simpler than before – IMO this is a step in the right direction.

Personally, I would prefer to throw our own ("API") version number out as well, but we can discuss that separately.

Finii commented:
I have a lot commits already done, on some machine, just need time to bat then into shape and push here. It will get much more simople than now already, iirc. Sorry for the stall and tnx for the comments.

_This is all so interrelated, also with Olaf's new template library, and how makeDdeb and co work, and it should maybe lead to some joint goal._

Fini -- heavily stressed by now also xmas stuff 😬

Finii commented:
This feels finished now, but I want to do some experiments still.

Finii commented:
This is (almost) like the clean-template-library now, and I guess good to pull.

(Rather here is one details that's missing over there).

Finii commented on LIBNO:
```diff
@@ -0,0 +1 @@
+LIBNO = 1.2.3
```
Well, I have no clue when GUL17 will be "activated" and it will get the then-current Doocs version number (automatically, by Tim) together with the proper git tag etc.

I would leave this to the maintainer who creates the first package.

Finii commented on README.md:
```diff
@@ -240,88 +197,5 @@ generate the HTML pages.
-        'D_'           false            gul17_18.11.7.p5.deb                  ->   libgul17.so.18.11.7
-        'D_'           true             gul17-18-11-7_18.11.7.p5.deb          ->   libgul17.so.18.11.7
+The versioning of is by an external (non-semantic) version number. The version
+resides in the LIBNO file.
```
In fact it should be `YYY.M.minor` (with leading zeros suppressed) or we will get a problem in 2100 when the version `00.1.0` will arrive. ;-)

Ok, not our problem ;-D

Finii commented on debian/control:
```diff
@@ -1,25 +1,23 @@
  libraries and programs.
 
-Package: @NAME-DEV@
+Package: dev-doocs-gul17
```
Well, we had these discussions with GUL14. Of course it would all work also in the ordinary Debian/Ubuntu way.
I kind of gave up here.
If we have this strange _version is in the name and you need to update your server to get a bugfixed library_ I guess we should put it into `/export/doocs`.

I do not really care (anymore). GUL seems to be Doocs-only anyhow, so why not follow `¯\_(ツ)_/¯`

Finii commented on src/meson.build:
```diff
@@ -14,95 +14,21 @@ libgul_src = files([
-endif
+lib_version_full = meson.project_version()
+lib_version_parts = meson.project_version().split('.')
+lib_version_num = lib_version_full # TODO
```
This is really a todo that I forgot about... working on it

Finii replied:
Well, this is about the so-versions.

Previously GUL had 'clean' versions unlike all the doocs libraries.
I would keep it that way, see the commit message.

Here a screenshot of typical versions we have in production.
Only sane exception seem to be GUL and libonline ;-D

![image](https://github.com/user-attachments/assets/fd0656f8-3cc4-4936-a9a8-88403446e0a7)

Wait, why are they marked with a star and printed yellow?

Oh my ... what a mess :-(
This is `doocsdev20`

![image](https://github.com/user-attachments/assets/3a0da8df-8f93-4f79-b9ac-75611eb26d52)

Anyhow, TODO done and pushed.

Finii replied:
Ah, here more how it was and how it now should be. This is before we dropped the API version completely; but the concept still holds and that had me marking it TODO

![image](https://github.com/user-attachments/assets/5d6ea8bc-6771-41c5-aeae-ac4808e88935)
